doc/scapy/bluetooth.rst

```diff
@@ -0,0 +1,426 @@
+  A daemon that provides access to Bluetooth devices over D-Bus.
+
+``bluetoothctl``
+  An interactive command-line program which interfaces with the ``bluetoothd``
+  over D-Bus.
+
+``hcitool``
+  A command-line program which interfaces directly with kernel interfaces.
+
+
+`Support for Classic Bluetooth in bluez is quite mature`__, however `BLE is
+under active development`__.
+
+__ http://www.bluez.org/profiles/
+
+__ https://git.kernel.org/pub/scm/bluetooth/bluez.git/tree/TODO
+
+First steps
+===========
+
+.. note::
+
+   You must run these examples as ``root``.  These have only been tested on
+   Linux, and require Scapy v2.4.3 or later.
+
+Verify Bluetooth device
+-----------------------
+
+Before doing anything else, you'll want to check that your Bluetooth device has
+actually been detected by the operating system:
+
+.. code-block:: console
+
+   $ hcitool dev
+   Devices:
+	   hci0	xx:xx:xx:xx:xx:xx
+
+
+.. _hci-open:
+
+Opening a HCI socket
+--------------------
+
+The first step in Scapy is to open a HCI socket to the underlying Bluetooth
+device:
+
+.. code-block:: pycon
+
+   >>> # Open a HCI socket to device hci0
+   >>> bt = BluetoothHCISocket(0)
+
+Send a control packet
+---------------------
+
+This packet contains no operation (ie: it does nothing), but it will test that
+you can communicate through the HCI device:
+
+.. code-block:: pycon
+
+   >>> # Send an empty command packet.
+   >>> bt.send(
+   ...   HCI_Hdr()/
+   ...   HCI_Command_Hdr())
+   4
+
+You should then get a response:
+
+.. code-block:: pycon
+
+   >>> # Receive a packet. This blocks until there is a packet.
+   >>> p = bt.recv()
+   
+   >>> p.show()
+   ###[ HCI header ]### 
+     type= Event
+   ###[ HCI Event header ]### 
+        code= 0xf
+        len= 4
+   ###[ Command Status ]### 
+           status= 1
+           number= 2
+           opcode= 0x0
+
+.. note::
+
+   On Scapy v2.4.2 and earlier, use ``bt.recv(MTU)``.
+
+.. _hci-non-blocking:
+
+Using non-blocking mode
+-----------------------
+
+By default, HCI sockets are opened in *blocking mode*, so your script will stall
+if there are no messages waiting -- the only way to get out of it is
+:kbd:`Control-c`.
+
+You can change this behaviour by switching to *non-blocking mode*:
+
+.. code-block:: pycon
+
+   >>> bt.ins.setblocking(False)
+
+In non-blocking mode, reading from the interface when there is no data will
+raise a ``BlockingIOError`` exception:
+
+.. code-block:: pycon
+
+   >>> bt.recv()
+   Traceback (most recent call last):
+     File "<console>", line 1, in <module>
+     File "scapy/layers/bluetooth.py", line 1085, in recv
+       return HCI_Hdr(self.ins.recv(x))
+   BlockingIOError: [Errno 11] Resource temporarily unavailable
+
+Non-blocking mode makes it easy to get all pending packets into a ``list``:
+
+.. code-block:: pycon
+
+   >>> pkts = []
+   >>> while True:
+   ...   try:
+   ...     pkts.append(bt.recv())
+   ...   except BlockingIOError:
+   ...     break
```

Comment:
Maybe it could be nice to add an example that uses `bt.sniff(orb=...)` or `bt.sr(...)`

Looking great so far !

Reply:
Hmm, `bt.sr` doesn't seem to actually work -- I should fix this.

Reply:
...and now fixed.